File: solutions/008/008.go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"strconv"
	"strings"
)

func sliceProduct(x []int) (product int) {
	product = 1
	for _, i := range x {
		if i == 0 {
			return 0
		}
		product *= i
	}
	return
}

func adjacentProduct(numbers []int, size int) int {
	largest := sliceProduct(numbers[:size])
	for i := size + 1; i < len(numbers); i++ {
		current := sliceProduct(numbers[i-size : i])
		if current > largest {
			largest = current
		}
	}
	return largest
}

func main() {
	file, _ := os.Open("./resources/digits.txt")
	defer file.Close()
	bs, _ := ioutil.ReadAll(file)
	numberString := strings.Replace(string(bs), "\n", "", -1)
	var numbers []int
	for _, n := range numberString {
		digit, _ := strconv.Atoi(string(n))
		numbers = append(numbers, digit)
	}

	fmt.Println(adjacentProduct(numbers, 13))
}

File: solutions/008/008.rs
use std::fs::File;
use std::io::Read;

fn main() {
    let mut numbers: [u8;1000] = [0;1000];

    let file = File::open("./resources/digits.txt");
    let mut i = 0;

    for byte in file.unwrap().bytes() {
        let b = byte.unwrap();
        if b >= 48 {
            numbers[i] = b - 48;
            i += 1;
        }
    }

    let size = 13;
    let mut largest: i64 = 0;

    for j in size..(1000-size) {
        let mut product: i64 = 1;
        for k in 0..size {
            product *= numbers[j+k] as i64;
        }
        if product > largest {
            largest = product;
        }
    }

    println!("{}", largest);
}
